### PR TITLE
Fixes for HostIPC tests to work when Docker has SELinux support enabled.

### PR DESCRIPTION
**Pull request by filbranden**
**What this PR does / why we need it**:

Fixes for HostIPC tests to work when Docker has SELinux support enabled.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

N/A

**Special notes for your reviewer**:

The core of the matter is to use `ipcs` from util-linux rather than the one from busybox. The typical SELinux policy has enough to allow Docker containers (running under svirt_lxc_net_t SELinux type) to access IPC information by reading the contents of the files under /proc/sysvipc/, but not by using the shmctl etc. syscalls.

The `ipcs` implementation in busybox will use `shmctl(0, SHM_INFO, ...)` to detect whether it can read IPC info (see source code [here](https://git.busybox.net/busybox/tree/util-linux/ipcs.c?h=1_28_0#n138)), while the one in util-linux will prefer to read from the /proc files directly if they are available (see source code [here](https://github.com/karelzak/util-linux/blob/v2.27.1/sys-utils/ipcutils.c#L108)).

It turns out the SELinux policy doesn't allow the shmctl syscalls in an unprivileged container, while access to it through the /proc interface is fine. (One could argue this is a bug in the SELinux policy, but getting it fixed on stable OSs is hard, and it's not that hard for us to test it with an util-linux `ipcs`, so I propose we do so.)

This PR also contains a refactor of the code setting IpcMode, since setting it in the "common options" function is misleading, as on containers other than the sandbox, it ends up always getting overwritten, so let's only set it to "host" in the Sandbox.

It also has a minor fix for the `ipcmk` call, since support for size suffix was only introduced in recent versions of it.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by filbranden** [comment]
For the record, the SELinux fix for this can be done using the following module (save it to a file named `docker_ipc.te`):

```selinux
module docker_ipc 1.0;

require {
        type unconfined_t;
        type kernel_t;
        type svirt_lxc_net_t;
        class shm { associate getattr unix_read };
        class system ipc_info;
}

#============= svirt_lxc_net_t ==============

allow svirt_lxc_net_t kernel_t:system ipc_info;
allow svirt_lxc_net_t unconfined_t:shm { associate getattr unix_read };
```

Compile and install it with:

```shell
$ checkmodule -M -m -o docker_ipc.mod docker_ipc.te
$ semodule_package -o docker_ipc.pp -m docker_ipc.mod
$ sudo semodule -i docker_ipc.pp
```

After that SELinux module is installed, `ipcs -m` will work inside a container that has IpcMode set to "container:\<id>".

**Comment by Random-Liu** [comment]
> The net result is more or less the same (since with the old approach they would essentially get "host" through the Pod sandbox container), but doing it explicitly this way has SELinux implications.

What selinux implication is that? I think it's good to know. AFAIK, both cri-containerd and dockershim have the same implementation - sandbox container has host namespace, other containers share sandbox container's namespace.

If this is a problem on SELinux, we may need to understand why, and whether it makes sense to make this change. And if it makes sense, I'll make the same change in cri-containerd. :)

**Comment by filbranden** [comment]
> And why only IPC namespace sharing has this problem?

The problem is actually with running `ipcs` command inside a container. By default, the SELinux policy prevents that from successfully working.

By default, this is what happens:

```
$ docker run busybox ipcs -m
kernel not configured for shared memory
```

So setting "--ipc host" in Docker makes the container run in a different SELinux context, which lets it run ipcs successfully:

```
$ docker run --ipc host busybox ipcs -m
------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status
```

Here's the difference in SELinux context with and without `--ipc host`:

```
$ docker run ubuntu ps -Z
LABEL                             PID TTY          TIME CMD
system_u:system_r:svirt_lxc_net_t:s0:c468,c779 1 ? 00:00:00 ps
$ docker run --ipc host ubuntu ps -Z
LABEL                             PID TTY          TIME CMD
system_u:system_r:spc_t:s0          1 ?        00:00:00 ps
```

Setting `--ipc` to point to another container doesn't trigger that SELinux context, so the `ipcs` command keeps failing.

See also my previous comment in this PR, there's some SELinux policy module that could be used to allow this.

NOTE: While testing this, I also noticed that this problem exists in busybox's `ipcs` only, it doesn't seem to affect the `ipcs` in e.g. Ubuntu. It's probably because the `ipcs` in busybox is doing additional work that Ubuntu's one is not doing.

> If this is a problem on SELinux, we may want to understand why, and whether it makes sense to make this change.

So I'm thinking this doesn't really make sense... I'm leaning towards:

1. Changing `modifyCommonNamespaceOptions`, so it doesn't set `hostConfig.IpcMode = namespaceModeHost` and move that to `modifySandboxNamespaceOptions`, since it's only relevant for the sandbox. That would clear up the confusion about it being overwritten later.

2. Either disable this test when SELinux is enabled, or maybe test it using a different container, such as "ubuntu" rather than "busybox". Do you think this would be acceptable?

Cheers!
Filipe

**Comment by filbranden** [comment]
@Random-Liu Please take a look. I also updated the PR description to reflect what we're doing here. I already pushed the `ipc-utils` container defined here to gcr.io (hope that's OK!)

Cheers,
Filipe

**Comment by filbranden** [comment]
Looks like a flake... so:

/test pull-kubernetes-e2e-kops-aws

File: pkg/kubelet/dockershim/security_context.go [hunk 1]
```diff
@@ -143,18 +145,17 @@ func modifyContainerNamespaceOptions(nsOpts *runtimeapi.NamespaceOption, podSand
...
-			hostConfig.IpcMode = namespaceModeHost
-		}
+	if nsOpts != nil && nsOpts.HostPid {
+		hostConfig.PidMode = namespaceModeHost
 	}
 }
 
 // modifyHostNetworkOptionForSandbox applies NetworkMode/UTSMode to sandbox's dockercontainer.HostConfig.
-func modifyHostNetworkOptionForSandbox(hostNetwork bool, network *knetwork.PluginManager, hc *dockercontainer.HostConfig) {
+func modifyHostNetworkOptionForSandbox(hostNetwork bool, hostIpc bool, network *knetwork.PluginManager, hc *dockercontainer.HostConfig) {
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/security_context.go]: Rename this function?

**Comment by Random-Liu** [comment]
LGTM with one nit.

File: pkg/kubelet/dockershim/security_context.go (hunk 1 above)
Comment (filbranden) [reply pkg/kubelet/dockershim/security_context.go]: Do you have any suggestions? Perhaps just `modifyHostOptionsForSandbox` (note "Options" plural.)

A few lines below, there's `modifyHostNetworkOptionForContainer` which also sets IpcMode and UTSMode, so I'm thinking for consistency we should rename that one too... Agreed?

**Review by filbranden — COMMENTED** [review]
Thanks for the review! I have a small follow up question.

**Comment by filbranden** [comment]
Let's try this again:

/test pull-kubernetes-e2e-kops-aws

File: pkg/kubelet/dockershim/security_context.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/kubelet/dockershim/security_context.go]: I prefer `modifyHostOptionsForSandbox`. :) Since you've passed in `hostIpc` here.

Comment (filbranden) [reply pkg/kubelet/dockershim/security_context.go]: Done! Also renamed modifyHostNetworkOptionForContainer -> modifyHostOptionsForContainer for consistency (that one also sets hostIpc, even though it doesn't take it as an argument...) PTAL.

**Comment by filbranden** [comment]
/assign @vishh 

Vish, can you please approve for OWNERS?

Thanks!
Filipe

**Comment by filbranden** [comment]
Retry...

/test pull-kubernetes-unit

**Comment by ixdy** [comment]
This PR appears to have broken the crossbuild:
```
I0131 05:29:12.462] # k8s.io/kubernetes/pkg/kubelet/dockershim
I0131 05:29:12.462] pkg/kubelet/dockershim/helpers_windows.go:52:3: undefined: modifyHostNetworkOptionForContainer
```

**Comment by cblecker** [comment]
I opened #59121 for the cross build failure

**Pull request merged**